src/utils/allocators.zig

```diff
@@ -2,6 +2,251 @@ const std = @import("std");
+        memory: std.ArrayList([]T),
+        // allocator used to alloc the memory blocks
+        memory_allocator: std.mem.Allocator,
+        capacity: u64,
+        // for thread safety
+        mux: std.Thread.Mutex = .{},
+
+        // NOTE: we use the global_index to support fast loading the state
+        pub const Record = struct {
+            is_free: bool,
+            buf: []T,
+            global_index: u64,
+            len: u64,
+            // NOTE: this is tracked for correct usage of collapse()
+            memory_index: u64,
+
+            /// NOTE: we dont want to re-write the buf memory (that exists in `memory` field) on save
+            pub const @"!bincode-config:buf": sig.bincode.FieldConfig([]T) = .{
+                .skip = true,
+                .default_value = &.{},
+            };
+        };
+        const Self = @This();
+
+        const AllocatorConfig = struct {
+            memory_allocator: std.mem.Allocator,
+            records_allocator: std.mem.Allocator,
+        };
+
+        pub fn init(allocator_config: AllocatorConfig) Self {
+            return .{
+                .records = std.ArrayList(Record).init(allocator_config.records_allocator),
+                .memory = std.ArrayList([]T).init(allocator_config.records_allocator),
+                .memory_allocator = allocator_config.memory_allocator,
+                .capacity = 0,
+            };
+        }
+
+        pub fn create(allocator_config: AllocatorConfig) !*Self {
+            const self = try allocator_config.records_allocator.create(Self);
+            self.* = Self.init(allocator_config);
+            return self;
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            for (self.memory.items) |block| {
+                self.memory_allocator.free(block);
+            }
+            self.memory.deinit();
+            self.records.deinit();
+        }
+
+        /// append a block of N elements to the manager
+        pub fn append(self: *Self, n: u64) !void {
+            if (n == 0) return;
+
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            const buf = try self.memory_allocator.alloc(T, n);
+            // NOTE: we do this here so bincode serialization can work correctly
+            // otherwise, we run into undefined memory which breaks bincode
+            @memset(buf, T.DEFAULT);
+
+            try self.records.append(.{
+                .is_free = true,
+                .buf = buf,
+                .global_index = self.capacity,
+                .len = buf.len,
+                .memory_index = self.memory.items.len,
+            });
+            try self.memory.append(buf);
+            self.capacity += buf.len;
+        }
+
+        pub fn alloc(self: *Self, n: u64) !struct { []T, u64 } {
```

Comment:
Similar to `append`, should this not also have a `if (n == 0) return;` check? if not it looks like it would attempt to allocate 0 sized buf, which is not that useful.

Reply:
32f5d8a0